widget/list.go

```diff
@@ -0,0 +1,415 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/driver/desktop"
+	"fyne.io/fyne/internal/widget"
+	"fyne.io/fyne/layout"
+	"fyne.io/fyne/theme"
+)
+
+// List is a widget that caches list items for performance and
+// lays the items out in a vertical direction inside of a scroller
+// List requires that all items are the same size
+type List struct {
+	BaseWidget
+
+	Length         func() int
+	CreateItem     func() fyne.CanvasObject
+	UpdateItem     func(index int, item fyne.CanvasObject)
+	OnItemSelected func(index int, item fyne.CanvasObject)
+
+	background      color.Color
+	offsetY         int
+	previousOffsetY int
+	selectedItem    *listItem
+	selectedIndex   int
+	itemMin         fyne.Size
+}
+
+// NewList creates and returns a list widget for displaying items in
+// a vertical layout with scrolling and caching for performance
+func NewList(length func() int, createItem func() fyne.CanvasObject, updateItem func(index int, item fyne.CanvasObject)) *List {
+	return &List{BaseWidget: BaseWidget{}, Length: length, CreateItem: createItem, UpdateItem: updateItem, selectedIndex: -1}
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (l *List) CreateRenderer() fyne.WidgetRenderer {
+	l.ExtendBaseWidget(l)
+
+	if l.itemMin.IsZero() && l.CreateItem != nil {
+		l.itemMin = newListItem(l.CreateItem(), nil).MinSize()
+	}
+	layout := fyne.NewContainerWithLayout(newListLayout(l))
+	layout.Resize(layout.MinSize())
+
+	scroller := NewVScrollContainer(layout)
+
+	scroller.onOffsetChanged = func() {
+		if l.offsetY == scroller.Offset.Y {
+			return
+		}
+		l.offsetY = scroller.Offset.Y
+		l.BaseWidget.Refresh()
+	}
+
+	objects := []fyne.CanvasObject{scroller}
+	return &listRenderer{BaseRenderer: widget.NewBaseRenderer(objects), list: l, scroller: scroller, layout: layout}
+}
+
+// MinSize returns the size that this widget should not shrink below
+func (l *List) MinSize() fyne.Size {
+	l.ExtendBaseWidget(l)
+
+	return l.BaseWidget.MinSize()
+}
+
+type listRenderer struct {
+	widget.BaseRenderer
+	list             *List
+	scroller         *ScrollContainer
+	layout           *fyne.Container
+	itemCache        *listPool
+	children         []fyne.CanvasObject
+	visibleItemCount int
+	firstItemIndex   int
+	lastItemIndex    int
+	size             fyne.Size
+}
+
+func (l *listRenderer) BackgroundColor() color.Color {
+	if l.list.background == nil {
+		return theme.BackgroundColor()
+	}
+
+	return l.list.background
+}
+
+func (l *listRenderer) Layout(size fyne.Size) {
+	if size != l.size {
```

Comment:
Is this needed? Layout should only be called when size has changed...

Reply:
layout is also called when the scroller offset has changed.

Reply:
I moved the length() == 0 quick exit to the top and added a cleanup in case the length of the list has changed from a positive count i.e. 100 to 0 to clean up the items.  As far as the rest... I'm not sure I follow?